chain/chain/src/chain.rs

```diff
@@ -1874,80 +1957,106 @@ impl Chain {
-        // TODO: uncomment this line when we are ready to move apply_chunks async
-        //rayon::spawn(move || {
-        let res = do_apply_chunks(work);
-        // we can safely unwrap here because error means the receiver is deallocated, which means
-        // the chain thread is shut down, and the node already crashed
-        sc.send((block_hash, res)).unwrap();
-        //});
+        rayon::spawn(move || {
+            let res = do_apply_chunks(work);
+            // we can safely unwrap here because error means the receiver is deallocated, which means
+            // the chain thread is shut down, and the node already crashed
+            sc.send((block_hash.clone(), res)).unwrap();
+            if let Err(_) = apply_chunks_done_marker.set(true) {
+                // This should never happen, if it does, it means there is a bug in our code.
+                error!(target:"chain", "apply chunks are called twice for block {:?}", block_hash);
+                debug_assert!(false);
+            }
+            apply_chunks_done_callback(block_hash);
+        });
     }
 
-    /// Check if there are any block that finished applying chunks. Run postprocessing on the block.
-    fn postprocess_ready_blocks(
+    /// Run postprocessing on this block, which stores the block on chain.
+    /// Check that if accepting the block unlocks any orphans in the orphan pool and start
+    /// the processing of those blocks.
+    fn postprocess_block(
         &mut self,
         me: &Option<AccountId>,
-        block_accepted: &mut Vec<AcceptedBlock>,
-    ) -> Result<Option<Tip>, Error> {
-        if let Ok((block_hash, apply_results)) = self.apply_chunks_receiver.try_recv() {
-            let (block, block_preprocess_info) =
-                self.blocks_in_processing.remove(&block_hash).expect(&format!(
-                    "block {:?} finished applying chunks but not in blocks_in_processing pool",
-                    block_hash
-                ));
-            let prev_head = self.store.head()?;
-            let mut chain_update = self.chain_update();
-            let provenance = block_preprocess_info.provenance.clone();
-            let new_head =
-                chain_update.postprocess_block(me, &block, block_preprocess_info, apply_results)?;
-            chain_update.chain_store_update.save_block_height_processed(block.header().height());
-            chain_update.commit()?;
+        block_hash: CryptoHash,
+        apply_results: Vec<Result<ApplyChunkResult, Error>>,
+        block_processing_artifacts: &mut BlockProcessingArtifact,
+        apply_chunks_done_callback: ApplyChunkCallback,
+    ) -> Result<AcceptedBlock, Error> {
+        let (block, block_preprocess_info) =
+            self.blocks_in_processing.remove(&block_hash).expect(&format!(
+                "block {:?} finished applying chunks but not in blocks_in_processing pool",
+                block_hash
+            ));
+        let _span = tracing::debug_span!(
```

Comment:
Agree that it doesn't matter here, but, just in case, tracing allows recording span additional values after the span is created: 

https://docs.rs/tracing/latest/tracing/span/struct.Span.html#method.record

One fun thing we use this for already is to record gas at the end of the span, such that a span contains both wall-clock and gas.